### PR TITLE
Codebase cleanup. Bug fixes

### PR DESCRIPTION
**Pull request by sebr72**
1. Avoid running testpdfA twice. Introduce new list to validate pdfs.
2. Fix bug stopping method from being transactional
3. Make inner classes static to avoid memory leaks
4. Improve logging messages
5. Improve performance using empty arrays. Also remove data race risk between the size and toArray calls
6. Remove dead code

### CODE REVIEW

File: core/src/main/java/org/mapfish/print/attribute/DataSourceAttribute.java [hunk 1]
```diff
@@ -113,14 +113,14 @@ public Map<String, Attribute> getAttributes() {
    */
   public void setAttributes(final Map<String, Attribute> attributes) {
     for (Map.Entry<String, Attribute> entry : attributes.entrySet()) {
-      Object attribute = entry.getValue();
-      if (!(attribute instanceof Attribute)) {
+      Attribute attribute = entry.getValue();
+      if (attribute == null) {
         final String msg =
-            "Attribute: '" + entry.getKey() + "' is not an attribute. It is a: " + attribute;
+            "Attribute: '" + entry.getKey() + "' is not an attribute. It is a: " + null;
```
Comment (sbrunner) [inline core/src/main/java/org/mapfish/print/attribute/DataSourceAttribute.java:119]:
```suggestion
            "Attribute: '" + entry.getKey() + "' is not an defined attribute.";
```

File: core/src/main/java/org/mapfish/print/config/Template.java [hunk 2]
```diff
@@ -131,14 +131,14 @@ public final Map<String, Attribute> getAttributes() {
    */
   public final void setAttributes(final Map<String, Attribute> attributes) {
     for (Map.Entry<String, Attribute> entry : attributes.entrySet()) {
-      Object attribute = entry.getValue();
-      if (!(attribute instanceof Attribute)) {
+      Attribute attribute = entry.getValue();
+      if (attribute == null) {
         final String msg =
-            "Attribute: '" + entry.getKey() + "' is not an attribute. It is a: " + attribute;
+            "Attribute: '" + entry.getKey() + "' is not an attribute. It is a: " + null;
```
Comment (sbrunner) [inline core/src/main/java/org/mapfish/print/config/Template.java:137]: Same as above

File: core/src/main/java/org/mapfish/print/config/Template.java [hunk 3]
```diff
@@ -168,9 +168,9 @@ public final void setProcessors(final List<Processor> processors) {
 
   private void assertProcessors(final List<Processor> processorsToCheck) {
     for (Processor entry : processorsToCheck) {
-      if (!(entry instanceof Processor)) {
-        final String msg = "Processor: " + entry + " is not a processor.";
-        LOGGER.error("Error setting the Attributes: {}", msg);
+      if (entry == null) {
+        final String msg = "Processor: " + null + " is not a processor.";
```
Comment (sbrunner) [inline core/src/main/java/org/mapfish/print/config/Template.java:172]:
```suggestion
        final String msg = "Processor: is missing.";
```

File: core/src/main/java/org/mapfish/print/http/MfClientHttpRequestFactoryImpl.java [hunk 4]
```diff
@@ -113,7 +113,7 @@ private static final class RandomizingDnsResolver extends SystemDefaultDnsResolv
     public InetAddress[] resolve(final String host) throws UnknownHostException {
       final List<InetAddress> list = Arrays.asList(super.resolve(host));
       Collections.shuffle(list);
-      return list.toArray(new InetAddress[list.size()]);
+      return list.toArray(new InetAddress[0]);
```
Comment (sbrunner) [inline core/src/main/java/org/mapfish/print/http/MfClientHttpRequestFactoryImpl.java:116]: Why this change?

File: core/src/main/java/org/mapfish/print/map/style/json/MapfishJsonStyleVersion2.java [hunk 5]
```diff
@@ -27,7 +27,7 @@
 public final class MapfishJsonStyleVersion2 {
   static final String JSON_SYMB = "symbolizers";
   private static final String JSON_TYPE = "type";
-  private static final Pattern VALUE_EXPR_PATTERN = Pattern.compile("\\$\\{([\\w\\d_-]+)\\}");
+  private static final Pattern VALUE_EXPR_PATTERN = Pattern.compile("\\$\\{([\\w_-]+)\\}");
```
Comment (sbrunner) [inline core/src/main/java/org/mapfish/print/map/style/json/MapfishJsonStyleVersion2.java:30]: Why remove '\\d'?

File: core/src/main/java/org/mapfish/print/processor/map/CreateMapPagesProcessor.java [hunk 6]
```diff
@@ -199,7 +199,7 @@ public final Output execute(final Input values, final ExecutionContext context)
     }
     LOGGER.info("Paging generate {} maps definitions.", mapList.size());
     DataSourceAttributeValue datasourceAttributes = new DataSourceAttributeValue();
-    datasourceAttributes.attributesValues = mapList.toArray(new Map[mapList.size()]);
+    datasourceAttributes.attributesValues = mapList.toArray(new Map[0]);
```
Comment (sbrunner) [inline core/src/main/java/org/mapfish/print/processor/map/CreateMapPagesProcessor.java:202]: Why this change?

File: core/src/main/java/org/mapfish/print/wrapper/multi/PMultiObject.java [hunk 7]
```diff
@@ -135,7 +135,7 @@ public final PObject optObject(final String key) {
     if (results.size() == 1) {
       return results.get(0);
     }
-    return new PMultiObject(results.toArray(new PObject[results.size()]));
+    return new PMultiObject(results.toArray(new PObject[0]));
```
Comment (sbrunner) [inline core/src/main/java/org/mapfish/print/wrapper/multi/PMultiObject.java:138]: ?

File: core/src/main/java/org/mapfish/print/http/MfClientHttpRequestFactoryImpl.java (hunk 4 above)
Comment (sebr72) [reply core/src/main/java/org/mapfish/print/http/MfClientHttpRequestFactoryImpl.java:116]: Qodana reported:
n older Java versions, using a pre-sized array was recommended, as the reflection call necessary to create an array of proper size was quite slow.
However, since late updates of OpenJDK 6, this call was intrinsified, making the performance of the empty array version the same, and sometimes even better, compared to the pre-sized version. Also, passing a pre-sized array is dangerous for a concurrent or synchronized collection as a data race is possible between the size and toArray calls. This may result in extra nulls at the end of the array if the collection was concurrently shrunk during the operation.

File: core/src/main/java/org/mapfish/print/processor/map/CreateMapPagesProcessor.java (hunk 6 above)
Comment (sebr72) [reply core/src/main/java/org/mapfish/print/processor/map/CreateMapPagesProcessor.java:202]: Same as above ;-)

File: core/src/main/java/org/mapfish/print/wrapper/multi/PMultiObject.java (hunk 7 above)
Comment (sebr72) [reply core/src/main/java/org/mapfish/print/wrapper/multi/PMultiObject.java:138]: Same as above

File: core/src/main/java/org/mapfish/print/map/style/json/MapfishJsonStyleVersion2.java (hunk 5 above)
Comment (sebr72) [reply core/src/main/java/org/mapfish/print/map/style/json/MapfishJsonStyleVersion2.java:30]: \d was redundant:
\w\d is the same as \w

**Review by sbrunner — APPROVED** [review]
